File: docproc/document_processing_system/report/expiring_documents/expiring_documents.js
frappe.query_reports["Expiring Documents"] = {
    "filters": [
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nOpen\nClosed",
            "default": "Open",
            "reqd": 0
        },
        {
            "fieldname": "bill_to",
            "label": __("Bill To"),
            "fieldtype": "Select",
            "options": "",
            "reqd": 0
        },
        {
            "fieldname": "expires_in",
            "label": __("Expires In"),
            "fieldtype": "Int",
            "default": 10,
            "reqd": 0
        },
        {
            "fieldname": "alert_type",
            "label": __("Document Type"),
            "fieldtype": "Select",
			"options": "\nBusiness\nPersonnel",
			"default": "",
            "reqd": 0
        }
    ],
    onload: function(report) {

       // Add a custom button to create quotation
        // report.page.add_inner_button(__('Create Quotation'), function() {
        //     openDialog(report);
        // }).removeClass('btn-default').addClass('btn-primary');


        // Fetch and set the bill_to options when the report loads
        generate_bill_to_options(report);

        // Add custom CSS
        $('<style type="text/css"> .red { color: red; } .maroon { color: maroon; } .yellow { color: yellow; } .blue { color: blue; } .orange { color: orange; } .cyan { color: cyan; } .green { color: green; } </style>').appendTo("head");
    },
    formatter: function (value, row, column, data, default_formatter) {
        if (column.fieldname == "status") {
            if (value == "Open") {
                value = "<span class='blue'>" + value + "</span>";
            } else if (value == "Follow-Up") {
                value = "<span class='orange'>" + value + "</span>";
            } else if (value == "Quotation") {
                value = "<span class='cyan'>" + value + "</span>";
            } else if (value == "Work-Order") {
                value = "<span class='green'>" + value + "</span>";
            } else if (value == "Cancelled") {
                value = "<span class='red'>" + value + "</span>";
            }
        }

        return default_formatter(value, row, column, data);
    }
};

function generate_bill_to_options(report) {
    frappe.call({
        method: 'docproc.document_processing_system.doctype.alert.alert.get_bill_to_values',
        callback: function(r) {
            if (r.message) {
                const bill_to_filter = report.get_filter('bill_to');
                bill_to_filter.df.options = '\n' + r.message.join('\n');
                bill_to_filter.refresh();
            }
        }
    });
}

// Function to open the dialog
function openDialog(report) {
    // Extract unique Bill To options from the report data
    let bill_to_options = [...new Set(report.data.map(row => row.bill_to))];

    const dialog = new frappe.ui.Dialog({
        title: 'Select Bill To and Documents',
        fields: [
            {
                label: 'Select Bill To',
                fieldname: 'bill_to',
                fieldtype: 'Select',
                options: bill_to_options,
                onchange: function() {
                    // Filter documents based on the selected Bill To
                    const selected_bill_to = this.value;
                    const filtered_documents = report.data.filter(row => row.bill_to === selected_bill_to);
                    const document_table = dialog.fields_dict.documents.df.data;
                    document_table.length = 0;  // Clear previous data
                    filtered_documents.forEach(row => {
                        document_table.push({
                            document: row.document_type,
                            document_id: row.document_id
                        });
                    });
                    dialog.fields_dict.documents.grid.refresh();
                }
            },
            {
                fieldtype: 'Table',
                fieldname: 'documents',
                label: 'Documents',
                fields: [
                    {
                        fieldtype: 'Data',
                        fieldname: 'document',
                        label: 'Document',
                        read_only: 1,
                        in_list_view: 1
                    },
                    {
                        fieldtype: 'Data',
                        fieldname: 'document_id',
                        label: 'Document ID',
                        read_only: 1,
                        in_list_view: 1
                    }
                ],
                data: [],
                get_data: function() {
                    return this.data;
                }
            }
        ],
        primary_action_label: 'Generate Quote',
        primary_action(values) {
            console.log(values);
            // LOGIC to create Quote
            dialog.hide();
        }
    });

    dialog.show();
}

File: docproc/document_processing_system/report/expiring_documents/expiring_documents.py
import frappe
from frappe import _


def execute(filters=None):
	if filters == None:
		filters = {}

	columns = get_columns()
	data = get_data(filters)

	charts = get_charts(data)[1]
	summary = get_summary(data)

	return columns, data, None, None, summary

def get_columns():
    return [
        {"label": _("Document"), "fieldname": "document_type", "fieldtype": "Data", "width": 150},
        {"label": _("Document ID"), "fieldname": "document_id", "fieldtype": "Data", "width": 150},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 100},
		{"label": _("Bill To"), "fieldname": "bill_to", "fieldtype": "Data", "width": 150},
        {"label": _("Date of Issue"), "fieldname": "date_of_issue", "fieldtype": "Date", "width": 100},
        {"label": _("Date of Expiry"), "fieldname": "date_of_expiry", "fieldtype": "Date", "width": 100},
        {"label": _("Expires In"), "fieldname": "expires_in", "fieldtype": "Int", "width": 100},
        {"label": _("Alert"), "fieldname": "alert_name", "fieldtype": "Link", "options": "Alert", "width": 150},
        {"label": _("Alert Type"), "fieldname": "alert_type", "fieldtype": "Data", "width": 100},
        {"label": _("Created On"), "fieldname": "creation", "fieldtype": "Date", "width": 100},
        {"label": _("Last Updated On"), "fieldname": "modified", "fieldtype": "Date", "width": 100}
    ]

def get_data(filters):
	conditions = get_conditions(filters)
	query = f"""
		SELECT
			ad.document_type,
			ad.document_id,
			ad.date_of_issue,
			ad.date_of_expiry,
			CAST(ad.expires_in AS INT) as expires_in,
			a.name as alert_name,
			a.alert_type,
			a.bill_to,
			a.status,
			a.creation,
			a.modified
		FROM
			`tabAlert` a
			JOIN `tabAlert Documents` ad ON a.name = ad.parent
		WHERE
			{conditions}
	"""
	data = frappe.db.sql(query, filters, as_dict=1)
	return data

def get_conditions(filters):
    conditions = ["1=1"]
    if filters.get("status"):
        conditions.append("a.status = %(status)s")
    if filters.get("bill_to"):
        conditions.append("a.bill_to = %(bill_to)s")
    if filters.get("expires_in"):
        conditions.append("CAST(ad.expires_in AS INT) <= %(expires_in)s")
    if filters.get("alert_type"):
        conditions.append("a.alert_type = %(alert_type)s")

    return " AND ".join(conditions)

def get_charts(data):
    document_types = {}
    for row in data:
        document_type = row["document_type"]
        if document_type:
            document_types[document_type] = document_types.get(document_type, 0) + 1
    
    chart1 = {
        "type": "pie",
        "title": "Document Types Distribution",
        "data": {
            "labels": list(document_types.keys()),
            "datasets": [{
                "values": list(document_types.values())
            }]
        }
    }
    
    bill_tos = {}
    for row in data:
        bill_to = row["bill_to"]
        if bill_to:
            bill_tos[bill_to] = bill_tos.get(bill_to, 0) + 1
    
    chart2 = {
        "type": "bar",
        "title": "Documents by Bill To",
        "data": {
            "labels": list(bill_tos.keys()),
            "datasets": [{
                "values": list(bill_tos.values())
            }]
        }
    }

    return [chart1, chart2]

def get_summary(data):
	total_documents = len(data)
	expiring_soon = len([d for d in data if 0 < int(d['expires_in']) <= 7])
	alert_type_distribution = {
		"Personnel": len([d for d in data if d['alert_type'] == "Personnel"]),
		"Business": len([d for d in data if d['alert_type'] == "Business"])
	}

	personnel_type = alert_type_distribution["Personnel"]
	business_type = alert_type_distribution["Business"]
	
	avg_time_to_expiry = sum([int(d['expires_in']) for d in data]) / total_documents if total_documents > 0 else 0
	avg_time_to_expiry = round(avg_time_to_expiry, 1)


	expire_in_7_days = f'<a href="/app/query-report/Expiring%20Documents?status=Open&expires_in=7" style="color:red">{expiring_soon}</a>'
	personnel_docs = f'<a href="/app/query-report/Expiring%20Documents?status=Open&alert_type=Personnel&expires_in=30" style="color:orange">{personnel_type}</a>'
	business_docs = f'<a href="/app/query-report/Expiring%20Documents?status=Open&alert_type=Business&expires_in=30" style="color:orange">{business_type}</a>'

	report_summary = [
		{"value": total_documents, "label": "Total Alert Docs", "datatype": "Data"},
		{"value": expire_in_7_days, "label": "Critical (7 days)", "datatype": "Data"},
		{"value": personnel_docs, "label": "Personnel Docs", "datatype": "Data"},
		{"value": business_docs, "label": "Business Docs", "datatype": "Data"},
		{"value": f'{avg_time_to_expiry} days', "label": "Average Time to Expiry", "datatype": "Data"},
	]

	return report_summary
